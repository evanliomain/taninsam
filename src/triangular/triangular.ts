/**
 * @module array=>array
 */
/**
 * Transform an array in a triangular array
 * [A, B, C, D] =>
 * [
 *  [A],
 *  [A, B],
 *  [A, B, C],
 *  [A, B, C, D]
 * ]
 * @return the function to apply on the array to do transform it to a triangular array
 * @example
 * ```
 * triangular<number>()([1, 2, 3]) // [[1], [1, 2], [1, 2, 3]]
 * ```
 * @example Using the chain
 * ```
 * chain([1, 2, 3])
 *   .chain(triangular<number>())
 *   .value() // [[1], [1, 2], [1, 2, 3]]
 * ```
 */
export function triangular<T>(): (
  array: ReadonlyArray<T>
) => ReadonlyArray<ReadonlyArray<T>> {
  return (array: ReadonlyArray<T>): ReadonlyArray<ReadonlyArray<T>> =>
    array.reduce((acc: ReadonlyArray<ReadonlyArray<T>>, current: T) => {
      if (0 === acc.length) {
        return [[current]];
      }
      return [...acc, [...acc[acc.length - 1], current]];
    }, []);
}
