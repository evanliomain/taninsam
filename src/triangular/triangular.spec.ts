import { triangular } from './triangular';

describe('triangular function', () => {
  const input = [1, 2, 3, 4, 5];

  test('is a pure function', () => {
    triangular()(input);
    expect(input).toMatchSnapshot();
  });

  test('[1, 2, 3, 4, 5] |> triangular === [[1], [1, 2], [1, 2, 3], [1, 2, 3, 4], [1, 2, 3, 4, 5]]', () => {
    expect(triangular()(input)).toEqual([
      [1],
      [1, 2],
      [1, 2, 3],
      [1, 2, 3, 4],
      [1, 2, 3, 4, 5]
    ]);
  });
});
