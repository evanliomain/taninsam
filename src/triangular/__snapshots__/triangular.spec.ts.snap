// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`triangular function > [] |> triangular === [] 1`] = `[]`;

exports[`triangular function > [1, 2, 3, 4, 5] |> triangular === [[1], [1, 2], [1, 2, 3], [1, 2, 3, 4], [1, 2, 3, 4, 5]] 1`] = `
[
  [
    1,
  ],
  [
    1,
    2,
  ],
  [
    1,
    2,
    3,
  ],
  [
    1,
    2,
    3,
    4,
  ],
  [
    1,
    2,
    3,
    4,
    5,
  ],
]
`;

exports[`triangular function > is a pure function 1`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;
